01_stratified_classification/nbconverted/lasso_range_gene_optimizers.py

```diff
@@ -0,0 +1,791 @@
+
+fig, axarr = plt.subplots(2, 1)
+
+param_q_map = perf_coefs_df.loc[:, ['lasso_param', 'nz_quantile']]
+
+# ax1 is the bin axis for liblinear
+ax1 = axarr[0]
+ax1.scatter(x=param_q_map.nz_quantile.astype(int).values,
+            y=[0] * param_q_map.shape[0])
+ax1.set_xlim(param_q_map.nz_quantile.astype(int).min() - 0.5,
+             param_q_map.nz_quantile.astype(int).max() + 0.5)
+ax1.set_xticks(param_q_map.nz_quantile.astype(int).unique())
+ax1.get_yaxis().set_visible(False)
+ax1.set_xlabel('Quantile')
+
+ax2 = ax1.twiny()
+param_vals = param_q_map.lasso_param.sort_values(ascending=True).astype(str)
+ax2.scatter(x=param_vals, y=[1] * param_q_map.shape[0])
+ax2.set_xlabel('LASSO parameter, liblinear', labelpad=10)
+
+def bins_to_coords(bins):
+    # TODO document
+    x = np.linspace(ax2.get_xlim()[0], ax2.get_xlim()[1], len(perf_coefs_df.nz_quantile.unique())+1)
+    # https://stackoverflow.com/a/23856065
+    x_mid = (x[1:] + x[:-1]) / 2
+    return {b: x_mid[int(b)-1] for b in bins}
+
+# iterate through all (param, bin) coordinates and use bin to index
+qs = param_q_map.nz_quantile.astype(int).sort_values(ascending=True).unique().tolist()
+b_to_c = bins_to_coords(qs)
+
+unique_param_vals = param_vals.astype(float).unique().tolist()
+print(unique_param_vals)
+for ix, row in ll_param_q_map.iterrows():
+    bin_ix = int(row.nz_quantile)
+    lasso_param_ix = unique_param_vals.index(row.lasso_param) 
+    ax2.plot([b_to_c[bin_ix], lasso_param_ix], [0, 1], 'bo:')
+ax1.set_ylim(-0.1, 1.1)
+ax1.set_yticks([0, 1])
+
+# ax3 is the bin axis for SGD
+ax3 = axarr[1]
+ax3.scatter(x=param_q_map.nz_quantile.astype(int).values,
+            y=[0] * param_q_map.shape[0])
+ax3.set_xlim(param_q_map.nz_quantile.astype(int).min() - 0.5,
+             param_q_map.nz_quantile.astype(int).max() + 0.5)
+ax3.set_xticks(param_q_map.nz_quantile.astype(int).unique())
+ax3.get_yaxis().set_visible(False)
+ax3.set_xlabel('Quantile')
+
+ax4 = ax3.twiny()
+param_vals = param_q_map.lasso_param.sort_values(ascending=True).astype(str)
+ax4.scatter(x=param_vals, y=[1] * param_q_map.shape[0])
+ax4.set_xlabel('LASSO parameter, SGD', labelpad=10)
+
+unique_param_vals = param_vals.astype(float).unique().tolist()
+print(unique_param_vals)
+for ix, row in sgd_param_q_map.iterrows():
+    bin_ix = int(row.nz_quantile)
+    lasso_param_ix = unique_param_vals.index(row.lasso_param) 
+    ax4.plot([b_to_c[bin_ix], lasso_param_ix], [0, 1], 'bo:')
+ax3.set_ylim(-0.1, 1.1)
+ax3.set_yticks([0, 1])
+
+plt.tight_layout()
+
+
+# In[20]:
+
+
+sns.set_style('ticks')
+
+ll_plot_df = (
+    perf_coefs_df[(perf_coefs_df.optimizer == 'liblinear')]
+      .sort_values(by=['nz_linear_bin'])
+      .reset_index(drop=True)
+)
+
+sgd_plot_df = (
+    perf_coefs_df[(perf_coefs_df.optimizer == 'sgd')]
+      .sort_values(by=['nz_linear_bin'])
+      .reset_index(drop=True)
+)
+
+plot_df = pd.concat((ll_plot_df, sgd_plot_df))
+
+with sns.plotting_context('notebook', font_scale=1.6):
+    g = sns.relplot(
```

Comment:
Okay, now I see that it's so you could have a standardized way to compare with the same directionality of penalty vs. number of coefficients. Probably need to add more documentation about that for future readers.